packages/server/src/models/Locker.ts

```diff
@@ -0,0 +1,34 @@
+import {EventEmitter} from 'node:events'
+
+export interface Locker {
+  lock(id: string): Promise<void>
+  unlock(id: string): Promise<void>
+}
+
+export class MemoryLocker implements Locker {
+  private lockEvent = new EventEmitter()
+  private locks: Record<string, boolean> = {}
+
+  isLocked(id: string) {
+    return Boolean(this.locks[id])
+  }
+
+  lock(id: string): Promise<void> {
+    if (!this.isLocked(id)) {
+      this.locks[id] = true
+      return Promise.resolve()
+    }
+
+    return new Promise((resolve) => {
+      this.lockEvent.once(`unlock:${id}`, () => {
+        this.locks[id] = true
+        resolve()
+      })
+    })
```

Comment:
On unlock event you need to check if lock is available. Only first pending lock can perform a lock. Others need to resubscribe to event.

```suggestion
    return new Promise((resolve) => {
      const callback = () => {
        if (!this.isLocked(id)) {
          this.locks[id] = true
          resolve()
        } else {
          this.lockEvent.once(`unlock:${id}`, callback)
        }
      }
      this.lockEvent.once(`unlock:${id}`, callback)
    })
```